`Development`: Explicitly disable git gpg signing to fix tests in environments with global gpg signing enabled

reschandreas opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
I could not run all the tests (around 184 failed) because on my system I have gpg signing of commits enabled globally. In my opinion, this should not affect the tests, so I added `setSign(false)` to all the commits in the broken tests. This results, again in my opinion and please correct me if I'm wrong, in more stable tests because it is less dependent on the git config of the developer. 
<!-- If it fixes an open issue, please link to the issue here. -->

### Description
<!-- Describe your changes in detail -->

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- git installation

1. set signing to true, execute tests on `develop`
2. observe around 184 tests failing
3. set signing to false, test again
4. all tests should pass
5. now pull my changes and test again with the flag true and false
6. both times, all tests should pass


### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- The line coverage must be above 90% for changes files and you must use extensive and useful assertions for server tests and expect statements for client tests. -->
<!-- Note: Use the table below and confirm in the last column that you have implemented extensive assertions for server tests and expect statements for client tests. -->
<!--       You can use `supporting_script/generate_code_cov_table/generate_code_cov_table.py` to automatically generate one from the corresponding Bamboo build plan artefacts. -->
<!--       Remove rows with only trivial changes from the table. -->
<!--
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ExerciseService.java | 85% | ✅                           |
| programming-exercise.component.ts | 95% | ✅              |
-->

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

krusche commented on src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java:
```diff
@@ -973,7 +974,7 @@ public void combineAllStudentCommits(Repository repository, ProgrammingExercise
             var name = optionalStudent.map(User::getName).orElse(artemisGitName);
             var email = optionalStudent.map(User::getEmail).orElse(artemisGitEmail);
-            studentGit.commit().setMessage("All student changes in one commit").setCommitter(name, email).call();
+            studentGit.commit().setSign(false).setMessage("All student changes in one commit").setCommitter(name, email).call();
```
I suggest to refactor this into a method so that we do not forget to set this in future usages.
The method could take the git object, and create a CommitCommand in which `setSign(false)` is invoked

reschandreas replied:
thank you for the feedback, I added a method to both GitService and GitUtilService

basak-akan commented on src/test/java/de/tum/in/www1/artemis/localvcci/LocalVCIntegrationTest.java:
```diff
@@ -223,7 +223,7 @@ void testUserCreatesNewBranch() throws Exception {
         Files.createFile(testFilePath);
         assignmentRepository.localGit.add().addFilepattern(".").call();
-        assignmentRepository.localGit.commit().setMessage("Add new file").call();
+        assignmentRepository.localGit.commit().setSign(false).setMessage("Add new file").call();
```
The newly extracted method isn't called in here.

reschandreas replied:
thanks, I changed it

basak-akan commented on src/test/java/de/tum/in/www1/artemis/localvcci/LocalVCLocalCITestService.java:
```diff
@@ -364,7 +364,7 @@ public String commitFile(Path localRepositoryFolder, Git localGit, String fileNa
         Files.createFile(testFilePath);
         localGit.add().addFilepattern(".").call();
-        RevCommit commit = localGit.commit().setMessage("Add " + fileName).call();
+        RevCommit commit = localGit.commit().setSign(false).setMessage("Add " + fileName).call();
```
The newly extracted function isn't called in here as well.

reschandreas replied:
Yes, because neither the GitService nor GitUtilService was used in this file, I wanted to avoid introducing one or another because I was unsure which one to use.

reschandreas replied:
I added the GitService as a dependency, now the method is used

basak-akan commented on src/test/java/de/tum/in/www1/artemis/util/GitUtilService.java:
```diff
@@ -286,4 +287,14 @@ public void writeEmptyJsonFileToPath(Path path) throws Exception {
+     * @param git Git Repository Object.
+     * @return CommitCommand with signing set to false.
+     */
+    public CommitCommand commit(Git git) {
```
I noticed the same method in two classes. Is there a specific reason for this? If they're doing the same thing, maybe we could put this in one of them.

reschandreas replied:
I found that the tests use GitService and GitUtilService also has issues when no Signing Service is available. So I added it there as well.

basak-akan reviewed (commented):
I've reviewed your recent changes and left some comments on your code. 
Also, I have a question.

reschandreas commented:
Thanks, @basak-akan, for the review! I hope I answered your questions.

JohannesStoehr commented on src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java:
```diff
@@ -639,10 +640,20 @@ public Repository getExistingCheckedOutRepositoryByLocalPath(@NotNull Path local
     }
 
+    /**
+     * Creates a CommitCommand and sets signing to false.
```
Can you extend the comment to include a reason for not signing a commit and/or explaining what the effect of changing setSign to false will have?

reschandreas replied:
Good Idea, done!

JohannesStoehr reviewed (changes requested):
Left one comment. Additionally I agree with @basak-akan that the duplicated code should be refactored

reschandreas commented:
Thanks! I removed the method from the GitUtilService and added GitService as an `@Autowired` dependency in every file where it was possible. I'm not that happy about that, I would rather change the commit method from the service to a static method so it can 1) be used without autowiring and 2) also be used in the LocalRepository class. What do you think?

pal03377 reviewed (changes requested):
I have my local GPG key set to the SSH format (which works for my Git installation: I can sign commits), but when executing the tests, I still get this error:

```
ProgrammingExerciseIntegrationJenkinsGitlabTest > importProgrammingExercise_noProgrammingLanguage_badRequest() FAILED
    java.lang.IllegalArgumentException: Invalid value: gpg.format=ssh
        at org.eclipse.jgit.lib.DefaultTypedConfigGetter.getEnum(DefaultTypedConfigGetter.java:105)
        at org.eclipse.jgit.lib.Config.getEnum(Config.java:449)
        at org.eclipse.jgit.lib.GpgConfig.<init>(GpgConfig.java:86)
        at org.eclipse.jgit.api.CommitCommand.processOptions(CommitCommand.java:662)
        at org.eclipse.jgit.api.CommitCommand.call(CommitCommand.java:189)
        at de.tum.in.www1.artemis.exercise.programmingexercise.ProgrammingExerciseIntegrationTestService.setup(ProgrammingExerciseIntegrationTestService.java:224)
        at de.tum.in.www1.artemis.exercise.programmingexercise.ProgrammingExerciseIntegrationJenkinsGitlabTest.initTestCase(ProgrammingExerciseIntegrationJenkinsGitlabTest.java:43)
```

I believe this has something to do with the version of Git being used? I'm unsure if this can even be fixed with this PR. When I run `git --version` locally, I get `git version 2.41.0`.

reschandreas commented:
I don't think that's an issue with the git version as I'm running the same one. I'll try to replicate the problem

reschandreas commented:
Ok @pal03377 you were absolutely right, the git version used by egit seems to be [2.10](https://mvnrepository.com/artifact/org.eclipse.jgit/org.eclipse.jgit.ssh.apache/6.6.0.202305301015-r) and signing with ssh key is only possible since [2.34](https://stackoverflow.com/a/74887438). 

I also think that this is out of scope for this PR, given also that egit is already at the latest version. So we can either wait for an egit version that supports ssh signing or recommend to either use gpg signing or none at all.

pal03377 reviewed (approved):
Tested again locally with a non-SSH GPG key.

I still get failing tests when running all of them locally (e.g. this one:
`java.io.FileNotFoundException: class path resource [templates/java/plain_gradle/test/] cannot be resolved to URL because it does not exist`),
but these tests also fail on develop without GPG enabled. 

So overall, this PR is an improvement 👍 Thanks!

JohannesStoehr reviewed (dismissed):
Approve code changes

Strohgelaender reviewed (dismissed):
looks good to me.

You could additinally add a new Rule to the `ArchitectureTest` class enforcing the usage of `gitService.commit()` instead of `commit()` directly. But that's non-blocking for me.

basak-akan reviewed (approved):
code LGTM after changes

DominikRemo reviewed (approved):
Server-tests succeed locally and independent of `global commit.gpgsign` setting

Also tested participation in programming exercises on ts1. Works as expected

bassner commented on src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java:
```diff
@@ -639,10 +640,22 @@ public Repository getExistingCheckedOutRepositoryByLocalPath(@NotNull Path local
+     * @param git Git Repository Object.
+     * @return CommitCommand with signing set to false.
+     */
+    public CommitCommand commit(Git git) {
```
Could this method just be static? If yes, then you would not need to add the service as dependency to new locations.

reschandreas replied:
It can yes, I changed it. Thanks!

bassner commented on src/test/java/de/tum/in/www1/artemis/util/LocalRepository.java:
```diff
@@ -85,7 +85,7 @@ public void configureRepos(String localRepoFileName, Path originRepositoryFolder
         Files.createFile(filePath);
         localGit.add().addFilepattern("test.txt").call();
-        localGit.commit().setMessage("Initial commit").call();
+        localGit.commit().setSign(false).setMessage("Initial commit").call();
```
Let's use the extracted method here as well

reschandreas replied:
Done

bassner reviewed (changes requested):
Hey, thank you for your contribution. Please check the inline comments below
